static/widgets/compiler-picker-popup.ts

```diff
@@ -0,0 +1,266 @@
+    favoritesContainer: JQuery<HTMLElement>;
+
+    groups: {value: string; label: string}[];
+    options: (CompilerInfo & {$groups: string[]})[];
+    langId: string;
+
+    isaFilters: string[];
+    categoryFilters: string[];
+
+    sifter: sifter.Sifter;
+    searchResults: ReturnType<sifter.Sifter['search']> | undefined;
+
+    constructor(private readonly compilerPicker: CompilerPicker) {
+        this.modal = $('#compiler-picker-modal').clone(true);
+        this.searchBar = this.modal.find('.compiler-search');
+        this.architectures = this.modal.find('.architectures');
+        this.compilerTypes = this.modal.find('.compiler-types');
+        this.compilersContainer = this.modal.find('.compilers-row');
+        this.resultsContainer = this.modal.find('.compilers');
+        this.favoritesContainer = this.modal.find('.favorites');
+
+        this.modal.on('shown.bs.modal', () => {
+            this.searchBar[0].focus();
+        });
+    }
+
+    setLang(groups: {value: string; label: string}[], options: (CompilerInfo & {$groups: string[]})[], langId: string) {
+        this.groups = groups;
+        this.options = options;
+        this.langId = langId;
+        this.setupFilters();
+        this.sifter = new sifter.Sifter(options, {
+            diacritics: false,
+        });
+    }
+
+    setupFilters() {
+        // get available instruction sets
+        const compilers = Object.values(this.compilerPicker.compilerService.getCompilersForLang(this.langId) ?? {});
+        // If instructionSet is '', just label it unknown
+        const instruction_sets = compilers.map(compiler => compiler.instructionSet || 'other');
+        this.architectures.empty();
+        this.architectures.append(
+            ...unique(instruction_sets)
+                .sort()
+                .map(isa => `<span class="architecture" data-value=${isa}>${isa}</span>`),
+        );
+        // get available compiler types
+        const compilerTypes = compilers.map(compiler => compiler.compilerCategories ?? ['other']).flat();
+        this.compilerTypes.empty();
+        this.compilerTypes.append(
+            ...unique(compilerTypes)
+                .sort()
+                .map(type => `<span class="compiler-type" data-value=${type}>${type}</span>`),
+        );
+
+        // search box
+        this.searchBar.on('input', () => {
+            const query = unwrapString(this.searchBar.val()).trim();
+            if (query === '') {
+                this.searchResults = undefined;
+            } else {
+                this.searchResults = this.sifter.search(query, {
+                    fields: ['name'],
+                    conjunction: 'and',
+                    sort: CompilerService.getSelectizerOrder(),
+                });
+            }
+            this.fillCompilers();
+        });
+
+        // isa filters
+        $(this.architectures)
+            .find('.architecture')
+            .on('click', e => {
+                e.preventDefault();
```

Comment:
Mayyybe extract this function and make a function that takes a JQuery object (either `isaFilters` or `categoryFilter`), just so the same code isn't duplicated. It's _just_ long enough to make me want to not do it twice, but I see why a reasonable person might be ok with the duplication.

Reply:
great suggestion, ty!